File: AH-UN Schedule API/src/models/shift.py
from sqlmodel import Relationship, SQLModel, Field
from datetime import datetime
import uuid

class ShiftBase(SQLModel):
    id: str
    username: str
    start: datetime
    end: datetime | None
    canceled: bool


class Shift(ShiftBase, table=True):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    username: str = Field(foreign_key="user.username", index=True)
    start: datetime = Field(index=True)
    end: datetime | None = Field(default=None)
    canceled: bool = Field(default=False)

    user: "User" = Relationship(back_populates="shifts")

    def serialize(self) -> "ShiftRead":
        s = self.model_dump()

        s["user"] = self.user

        return ShiftRead(**s)


class ShiftRead(ShiftBase):
    user: "UserRead"


class ShiftUpdate(SQLModel):
    start: datetime | None = Field(default=None)
    end: datetime | None = Field(default=None)
    canceled: bool | None = Field(default=None)


from src.models.user import User, UserRead

File: AH-UN Schedule Android/app/src/main/java/dev/zoutigewolf/ahun_schedule/api/models/Shift.kt
package dev.zoutigewolf.ahun_schedule.api.models

import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Shift(
    @SerialName("id") val id: String,
    @SerialName("username") val username: String,
    @SerialName("start") var start: LocalDateTime,
    @SerialName("end") var end: LocalDateTime?,
    @SerialName("canceled") var canceled: Boolean,
    @SerialName("user") val user: User,
)

@Serializable
data class ShiftUpdate(
    @SerialName("start") var start: LocalDateTime?,
    @SerialName("end") var end: LocalDateTime?,
    @SerialName("canceled") var canceled: Boolean?,
)

fun Shift.toUpdate(): ShiftUpdate {
    return ShiftUpdate(
        start = this.start,
        end = this.end,
        canceled = this.canceled
    )
}